src/libfetchers/git.cc

```diff
@@ -4,16 +4,94 @@
 #include "tarfile.hh"
 #include "store-api.hh"
 #include "url-parts.hh"
+#include "util.hh"
 
 #include <sys/time.h>
 
 using namespace std::string_literals;
 
 namespace nix::fetchers {
 
-static std::string readHead(const Path & path)
+// Returns the name of the HEAD branch.
+//
+// Returns the head branch name as reported by git ls-remote, e.g.,
+// if ls-remote returns the output below, "main" is returned.
+//
+//   351037...    HEAD
+//   271366...    refs/heads/my-feature
+//   351037...    refs/heads/main
+static std::optional<std::string> readHead(const Path & path)
 {
-    return chomp(runProgram("git", true, { "-C", path, "rev-parse", "--abbrev-ref", "HEAD" }));
+    RunOptions opts("git", {"ls-remote", path});
+    std::pair<int, std::string> s = runProgram(opts);
+    auto& [exit_code, output] = s;
+    if (exit_code != 0) {
+        return std::nullopt;
+    }
+
+    const auto& tokens = tokenizeString<std::vector<std::string>>(output);
+    auto it = tokens.cbegin();
+    std::string headRev;
+    while (true) {
+        if (it == tokens.cend()) { break; }
+        const std::string_view rev = *it++;
+        if (it == tokens.cend()) { break; }
+        std::string_view ref = *it++;
+
+        if (ref == "HEAD") {
+            headRev = rev;
+        } else if (rev == headRev) {
+            constexpr std::string_view headsPath = "refs/heads/";
+            if (hasPrefix(ref, headsPath)) {
+                ref.remove_prefix(headsPath.size());
+            }
+            debug("resolved HEAD ref '%s' for repo '%s'", ref, path);
+            return std::string(ref);
+        }
+    }
+
+    return std::nullopt;
+}
+
+static std::optional<std::string> readHeadCached(std::string actualUrl)
+{
+    // Append something in front of the url to prevent collision with the
+    // cached repo.
+    Path cachePath = getCacheDir() + "/nix/gitv3/" + hashString(htSHA256, "<ref>|" + actualUrl).to_string(Base32, false);
```

Comment:
If I'm not mistaken, it is possible to get the remote `HEAD` by reading the `HEAD` file in the cached repository. Is it possible to do it here?

Reply:
Added a helper function for cache path.

`HEAD` is not currently updated in the cached repo (it is initialized to `__nix_dummy_branch`). I think we could use it instead of the cache file since it's not used for any other purpose, but it seems more straightforward to just write it to a separate file. Let me know if you still prefer using `HEAD`.

Reply:
I personally think that one cache dir is easier to maintain than two, but I do not feel that strongly about it. If you feel this is more straightforward, I won't bother you ^^

Reply:
On a side note the `getCachePath` function can also be used on line 418 of `git.cc`.

Reply:
Updated the pull request to maintain a single git repo as the cache, and storing the HEAD as a symbolic ref within the cache. This complicated the logic in `fetch()` slightly, so I also added a commit with some refactoring which I think makes this easier to understand overall.